packages/slate-react/src/plugins/android.js

```diff
@@ -0,0 +1,624 @@
+import Debug from 'debug'
```

Comment:
Why is this implemented as a separate plugin? It seems like it's going to make it really hard to maintain, since most fixes/changes won't be kept in sync? And it also means that we don't get any of the IME fixes that Android made in other platforms?

Reply:
So last year when I took a stab at this, I tried implementing Android as part of the regular `before` and `after` plugins; however, I found it difficult to implement because the changes were almost always to fix Android specific issues and they almost always had side effects that affected other browsers. I ended up having very low confidence that I didn't break things in other editors, and a very high rate of Android specific code mixed in and making it difficult to reason about the code for other browsers (lots of "if Android do this else do that statements").

Moving Android into a separate plugin is one reason development went so much faster this time. I didn't have to worry about regressions in other browsers.

However, that said, my hope is that we will see what's common between the different editors and the way they handle compositions and we could perhaps merge them together... possibly. But I can't say that with high confidence because most of the problems seem specific to Android. I feel a higher confidence level in abstracting compositions into a pluggable composition library that we could customize for each browser.

An example of Android having specific changes that don't affect other browsers is that in the normal Slate `before` plugin, there is some code in there that says something like if a composition ends but a new composition starts again then we should treat it as if the composition hasn't ended yet. That logic doesn't work in Android so I created an `isStrictComposing` in addition to the `isComposing` flag. So now we had two separate but similar concepts of `isComposing`. But later, we have to split code on whether to use isComposing or isStrictComposing depending on the browsers. Also as I dove deeper this time around, I realized that Android will actually do a `compositionEnd` without having a corresponding `compositionStart` which was a kind of implied composition. There is no corresponding logic in other browsers that I'm aware of. These situations had to be handled differently. Furthermore, there is just a lot of really convoluted logic that figures out intent through the combination of events and not a single event. This is in my code which I was planning to abstract into a Combo object.

I can see how looking at my code one might think that there must be an easier way. There is no such complexity in any of the other browsers. That was my feeling at first as well. I will continue to simplify and abstract the code to make it more understandable. I'm fairly confident though that at this point in time, a separate Android plugin makes the most sense and is, in fact, easier rather than harder to keep in sync by virtue of the fact that its easier to understand the code by (IMO) an order of magnitude.

You asked about IME fixes in other platforms as well. I do feel like the IME fixes for Android can be implemented into the rest of Slate but perhaps in another step; for example, there are already a lot of nice abstracted pieces that we can move into the regular handling that wasn't abstracted before. My first step was to get something working for Android though.